attributes.go

```diff
@@ -0,0 +1,256 @@
+
+	/* CKA_ECDSA_PARAMS is deprecated in v2.11,
+	 * CKA_EC_PARAMS is preferred. */
+	CkaEcdsaParams = AttributeType(0x00000180)
+	CkaEcParams    = AttributeType(0x00000180)
+
+	CkaEcPoint = AttributeType(0x00000181)
+
+	/* CKA_SECONDARY_AUTH, CKA_AUTH_PIN_FLAGS,
+	 * are new for v2.10. Deprecated in v2.11 and onwards. */
+	CkaSecondaryAuth = AttributeType(0x00000200) /* Deprecated */
+	CkaAuthPinFlags  = AttributeType(0x00000201) /* Deprecated */
+
+	CkaAlwaysAuthenticate = AttributeType(0x00000202)
+
+	CkaWrapWithTrusted = AttributeType(0x00000210)
+
+	ckfArrayAttribute = AttributeType(0x40000000)
+
+	CkaWrapTemplate   = AttributeType(ckfArrayAttribute | AttributeType(0x00000211))
+	CkaUnwrapTemplate = AttributeType(ckfArrayAttribute | AttributeType(0x00000212))
+
+	CkaOtpFormat               = AttributeType(0x00000220)
+	CkaOtpLength               = AttributeType(0x00000221)
+	CkaOtpTimeInterval         = AttributeType(0x00000222)
+	CkaOtpUserFriendlyMode     = AttributeType(0x00000223)
+	CkaOtpChallengeRequirement = AttributeType(0x00000224)
+	CkaOtpTimeRequirement      = AttributeType(0x00000225)
+	CkaOtpCounterRequirement   = AttributeType(0x00000226)
+	CkaOtpPinRequirement       = AttributeType(0x00000227)
+	CkaOtpCounter              = AttributeType(0x0000022E)
+	CkaOtpTime                 = AttributeType(0x0000022F)
+	CkaOtpUserIdentifier       = AttributeType(0x0000022A)
+	CkaOtpServiceIdentifier    = AttributeType(0x0000022B)
+	CkaOtpServiceLogo          = AttributeType(0x0000022C)
+	CkaOtpServiceLogoType      = AttributeType(0x0000022D)
+
+	CkaGostr3410Params = AttributeType(0x00000250)
+	CkaGostr3411Params = AttributeType(0x00000251)
+	CkaGost28147Params = AttributeType(0x00000252)
+
+	CkaHwFeatureType = AttributeType(0x00000300)
+	CkaResetOnInit   = AttributeType(0x00000301)
+	CkaHasReset      = AttributeType(0x00000302)
+
+	CkaPixelX                 = AttributeType(0x00000400)
+	CkaPixelY                 = AttributeType(0x00000401)
+	CkaResolution             = AttributeType(0x00000402)
+	CkaCharRows               = AttributeType(0x00000403)
+	CkaCharColumns            = AttributeType(0x00000404)
+	CkaColor                  = AttributeType(0x00000405)
+	CkaBitsPerPixel           = AttributeType(0x00000406)
+	CkaCharSets               = AttributeType(0x00000480)
+	CkaEncodingMethods        = AttributeType(0x00000481)
+	CkaMimeTypes              = AttributeType(0x00000482)
+	CkaMechanismType          = AttributeType(0x00000500)
+	CkaRequiredCmsAttributes  = AttributeType(0x00000501)
+	CkaDefatCmsAttributes     = AttributeType(0x00000502)
+	CkaSupportedCmsAttributes = AttributeType(0x00000503)
+	CkaAllowedMechanisms      = AttributeType(ckfArrayAttribute | AttributeType(0x00000600))
+)
+
+// An Attribute represents a PKCS#11 CK_ATTRIBUTE type.
+type Attribute = pkcs11.Attribute
+
+// An AttributeSet groups together operations that are common for a slice of Attributes
+type AttributeSet struct {
```

Comment:
We could just make this a map, for simplicity.

E.g.

```go
type AttributeSet map[AttributeType]*Attribute

func NewAttributeSet() AttributeSet {
	return make(AttributeSet)
}
```

The receiver functions would need to drop from `*AttributeSet` to `AttributeSet`. What do you think?

Reply:
Yeah this makes sense and actually makes the garbage collection faster since it does not have to track all of the pointers being passed around.